x/sequencer/keeper/unbond.go

```diff
@@ -26,158 +49,130 @@ func (k Keeper) UnbondAllMatureSequencers(ctx sdk.Context, currTime time.Time) {
-	seq, found := k.GetSequencer(ctx, seqAddr)
-	if !found {
-		return types.ErrUnknownSequencer
+func (k Keeper) reduceSequencerBond(ctx sdk.Context, seq *types.Sequencer, amt sdk.Coins, burn bool) error {
+	if amt.IsZero() {
+		return nil
 	}
-
-	if seq.Status == types.Unbonded {
+	if !seq.Tokens.IsAllGTE(amt) {
 		return errorsmod.Wrapf(
-			types.ErrInvalidSequencerStatus,
-			"sequencer status is already unbonded",
+			types.ErrInsufficientBond,
+			"insufficient bond for sequencer: %s", seq.Address,
 		)
 	}
-
-	oldStatus := seq.Status
-
-	seqTokens := seq.Tokens
-	if !seqTokens.Empty() {
-		seqAcc, err := sdk.AccAddressFromBech32(seq.Address)
+	if burn {
+		err := k.bankKeeper.BurnCoins(ctx, types.ModuleName, amt)
 		if err != nil {
 			return err
 		}
-
-		err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, seqAcc, seqTokens)
+	} else {
+		// refund
+		seqAcc := sdk.MustAccAddressFromBech32(seq.Address)
+		err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, seqAcc, amt)
 		if err != nil {
 			return err
 		}
-	} else {
-		k.Logger(ctx).Error("sequencer has no tokens to unbond", "sequencer", seq.Address)
 	}
 
-	// set the status to unbonded and remove from the unbonding queue if needed
-	seq.Status = types.Unbonded
-	seq.Proposer = false
-	seq.Tokens = sdk.Coins{}
-
-	k.UpdateSequencer(ctx, seq, oldStatus)
-
-	if oldStatus == types.Unbonding {
-		k.removeUnbondingSequencer(ctx, seq)
-	} else {
-		// in case the sequencer is currently reducing its bond, then we need to remove it from the decreasing bond queue
-		// all the tokens are returned, so we don't need to reduce the bond anymore
-		if bondReductions := k.getSequencerDecreasingBonds(ctx, seq.Address); len(bondReductions) > 0 {
-			for _, bondReduce := range bondReductions {
-				k.removeDecreasingBondQueue(ctx, bondReduce)
-			}
-		}
-	}
+	seq.Tokens = seq.Tokens.Sub(amt...)
+	return nil
+}
 
-	ctx.EventManager().EmitEvent(
-		sdk.NewEvent(
-			types.EventTypeUnbonded,
-			sdk.NewAttribute(types.AttributeKeySequencer, seqAddr),
-			sdk.NewAttribute(types.AttributeKeyBond, seqTokens.String()),
-		),
-	)
+func (k Keeper) unbondSequencerAndJail(ctx sdk.Context, seqAddr string) error {
+	return k.unbond(ctx, seqAddr, true)
+}
 
-	return nil
+func (k Keeper) unbondSequencer(ctx sdk.Context, seqAddr string) error {
+	return k.unbond(ctx, seqAddr, false)
 }
 
-// unbondUnbondingSequencer unbonds a sequencer that currently unbonding
-func (k Keeper) unbondUnbondingSequencer(ctx sdk.Context, seqAddr string) error {
+func (k Keeper) unbond(ctx sdk.Context, seqAddr string, jail bool) error {
 	seq, found := k.GetSequencer(ctx, seqAddr)
 	if !found {
 		return types.ErrUnknownSequencer
 	}
 
-	if seq.Status != types.Unbonding {
+	if seq.Status == types.Unbonded {
 		return errorsmod.Wrapf(
 			types.ErrInvalidSequencerStatus,
-			"sequencer status is not unbonding: got %s",
-			seq.Status.String(),
+			"sequencer status is already unbonded",
```

Comment:
need to document when this can happen
if it can't happen, needs to be a panic
in any case needs docs
we can just have all this defensive code it becomes impossible to understand anything

Reply:
panic is bad practice. bugs can happen.
not sure what's wrong with it

Reply:
it's dead code basically